api/dataset.go

```diff
@@ -176,66 +265,103 @@ func (api *DatasetAPI) addDataset(w http.ResponseWriter, r *http.Request) {
 	defer r.Body.Close()
 
 	datasetID := dataset.ID
-	dataset.Links.Self = "/datasets/" + datasetID
-	dataset.Links.Editions = "/datasets/" + datasetID + "/editions"
-	update := bson.M{
-		"$set": dataset,
-		"$setOnInsert": bson.M{
-			"updated_at": time.Now(),
-		},
+	dataset.Links.Self.Href = api.host + "/datasets/" + datasetID
+	dataset.Links.Editions.Href = api.host + "/datasets/" + datasetID + "/editions"
+	dataset.LastUpdated = time.Now()
+
+	var datasetDoc *models.DatasetUpdate
+
+	datasetDoc = &models.DatasetUpdate{
+		ID:   datasetID,
+		Next: dataset,
 	}
 
-	if err := api.dataStore.Backend.UpsertDataset(datasetID, update); err != nil {
+	if err := api.dataStore.Backend.UpsertDataset(datasetID, bson.M{"$set": datasetDoc}); err != nil {
 		log.ErrorR(r, err, nil)
 		w.WriteHeader(http.StatusInternalServerError)
 		return
 	}
 
+	bytes, err := json.Marshal(datasetDoc)
+	if err != nil {
+		log.Error(err, log.Data{"new_dataset": datasetID})
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
 	setJSONContentType(w)
 	w.WriteHeader(http.StatusCreated)
+	_, err = w.Write(bytes)
+	if err != nil {
+		log.Error(err, log.Data{"dataset_id": datasetID})
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+	}
 	log.Debug("upsert dataset", log.Data{"dataset_id": datasetID})
 }
 
 func (api *DatasetAPI) addEdition(w http.ResponseWriter, r *http.Request) {
 	vars := mux.Vars(r)
 	datasetID := vars["id"]
-	editionID := vars["edition"]
+	edition := vars["edition"]
+	if r.Header.Get(internalToken) != api.internalToken {
+		w.WriteHeader(http.StatusUnauthorized)
+		return
+	}
 
-	edition, err := models.CreateEdition(r.Body)
+	// Check if edition already exists and if it has been published return a status of Forbidden
+	currentEdition, err := api.dataStore.Backend.GetEdition(bson.M{"links.dataset.id": datasetID, "edition": edition})
+	if err != nil {
+		if err != api_errors.EditionNotFound {
+			log.Error(err, log.Data{"dataset_id": datasetID, "edition": edition})
+			handleErrorType(err, w)
+			return
+		}
+	} else {
```

Comment:
So i do need the else line so that the application does not complain when it tries the second if statement and cannot find a document to check its status. 

```
// handle error
if err != nil {
               // in the case of an error, check it was not a `not found error`
		if err != api_errors.EditionNotFound {
			log.Error(err, log.Data{"dataset_id": datasetID, "edition": edition})
			handleErrorType(err, w)
			return
		}
// if there are no errors (including `not found error`) and document is found
} else {
               // check if the state is set to published already
		if currentEdition.State == publishedState {
			w.WriteHeader(http.StatusForbidden)
			return
		}
}
```

Reply:
Ah ok I hadn't clocked the not found distinction. youve even put a comment there - ignore me!